### PR TITLE
crimson/osd: clear data_included if fiemap fails

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
@xxhdx1985126  could you help review this change ?

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -815,23 +815,16 @@ PG::get_or_load_head_obc(hobject_t oid)
           std::pair<crimson::osd::ObjectContextRef, bool>>
       {
 	logger().debug(
-	  "{}: loaded obs {} for {}",
-	  __func__,
-	  md->os.oi,
-	  oid);
+	  "get_or_load_head_obc: loaded obs {} for {}", md->os.oi, oid);
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc:818]: Yeah, the multi-line debugs were introducing too much visual clutter.

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 2]
```diff
@@ -516,14 +484,57 @@ seastar::future<ObjectRecoveryProgress> ReplicatedRecoveryBackend::build_push_op
...
+    return seastar::make_ready_future<uint64_t>(offset);
+  }
+  // 1. get the extents in the interested range
+  return backend->fiemap(coll, ghobject_t{oid},
+                         0, copy_subset.range_end()).then_wrapped(
+    [=,extents=copy_subset](auto&& f) mutable {
+    try {
+      interval_set<uint64_t> fiemap_included(f.get0());
+      extents.intersection_of(fiemap_included);
+    } catch (std::exception &) {
```
Comment (rzarzynski) [inline src/crimson/osd/replicated_recovery_backend.cc:510]: Do we expect an exception solely from `seastar::future::get0()` (which boils down to [`future_state::take()`](https://github.com/ceph/seastar/blob/master/include/seastar/core/future.hh#L568)) or from `intersection_of()` as well?

`CyanStore::fiemap()` throws `std::runtime_error` only:

```cpp
seastar::future<std::map<uint64_t, uint64_t>>
CyanStore::fiemap(
    CollectionRef ch,
    const ghobject_t& oid,
    uint64_t off,
    uint64_t len)
{
  auto c = static_cast<Collection*>(ch.get());

  ObjectRef o = c->get_object(oid);
  if (!o) {
    throw std::runtime_error(fmt::format("object does not exist: {}", oid));
  }
```

Maybe it's the time to errorate `fiemap()`? If so, then rather in a separated PR. Anyway, this is not a blocker.

Comment (tchaikov) [reply src/crimson/osd/replicated_recovery_backend.cc:510]: > Do we expect an exception solely from `seastar::future::get0()` (which boils down to [`future_state::take()`](https://github.com/ceph/seastar/blob/master/include/seastar/core/future.hh#L568)) or from `intersection_of()` as well?
> 

just from the former. to be specific, the `fiemap()` call.

> Maybe it's the time to errorate `fiemap()`? If so, then rather in a separated PR. Anyway, this is not a blocker.

agreed. we should not catch **all** exceptions. i will come up with a follow-up PR to errorate `fiemap()`.

Comment (tchaikov) [reply src/crimson/osd/replicated_recovery_backend.cc:510]: the reason i put `extents.intersection_of(fiemap_included);` in the `try` block is just to limit the lexical scope of `fiemap_included`. as what is interesting is `extents` here. i revised the comment and the `try` block a little bit. hopefully it's less confusing now.

**Pull request merged**